### PR TITLE
nearly optimal message pool seletion

### PR DESCRIPTION
**Pull request by vyzo**
Closes https://github.com/filecoin-project/lotus/issues/2948

### CODE REVIEW

File: chain/messagepool/block_proba_test.go [hunk 1]
```diff
@@ -0,0 +1,8 @@
+package messagepool
+
+import "testing"
+
+func TestBlockProbability(t *testing.T) {
+	mp := &MessagePool{}
+	t.Logf("%+v\n", mp.blockProbabilities(1-0.15))
```
Comment (whyrusleeping) [inline chain/messagepool/block_proba_test.go:7]: not much of a test

Comment (vyzo) [reply chain/messagepool/block_proba_test.go:7]: I think kuba wanted to eyeball it, but we can probably turn it into a real test :)

Comment (vyzo) [reply chain/messagepool/block_proba_test.go]: I turned it into an actual test.

File: miner/miner.go [hunk 2]
```diff
@@ -390,8 +392,14 @@ func (m *Miner) mineOne(ctx context.Context, base *MiningBase) (*types.BlockMsg,
 		return nil, xerrors.Errorf("failed to compute winning post proof: %w", err)
 	}
 
+	ticketHash := blake2b.Sum256(ticket.VRFProof)
+	ticketNum := types.BigFromBytes(ticketHash[:]).Int
+	ticketDenu := big.NewInt(1)
```
Comment (whyrusleeping) [inline miner/miner.go:397]: ideally the 'ticket to fraction' logic can be pulled into a separate function somewhere

File: node/impl/full/mpool.go [hunk 3]
```diff
@@ -40,13 +37,14 @@ func (a *MpoolAPI) MpoolSetConfig(ctx context.Context, cfg *types.MpoolConfig) e
...
 
-func (a *MpoolAPI) MpoolSelect(ctx context.Context, tsk types.TipSetKey) ([]*types.SignedMessage, error) {
+func (a *MpoolAPI) MpoolSelect(ctx context.Context, tsk types.TipSetKey, ticketQuality float64) ([]*types.SignedMessage, error) {
 	ts, err := a.Chain.GetTipSetFromKey(tsk)
 	if err != nil {
 		return nil, xerrors.Errorf("loading tipset %s: %w", tsk, err)
 	}
 
-	return a.Mpool.SelectMessages(ts)
+	// TODO FIXME compute (or pass in) the actual ticket quality!
```
Comment (whyrusleeping) [inline node/impl/full/mpool.go:46]: is this TODO still relevant?

Comment (Kubuxu) [reply node/impl/full/mpool.go:46]: No

File: chain/messagepool/selection.go [hunk 4]
```diff
@@ -548,7 +810,13 @@ func (mc *msgChain) Trim(gasLimit int64, mp *MessagePool, baseFee types.BigInt,
 		mc.gasReward = new(big.Int).Sub(mc.gasReward, gasReward)
 		mc.gasLimit -= mc.msgs[i].Message.GasLimit
 		if mc.gasLimit > 0 {
+			bp := 1.0
+			if mc.effPerf != 0 {
+				bp = mc.effPerf / mc.gasPerf
```
Comment (Kubuxu) [inline chain/messagepool/selection.go:815]: Uhh, missed that.
if `mc.gasPerf == 0` then bp is `Inf` so effPerf is `Inf`

Comment (Kubuxu) [reply chain/messagepool/selection.go:815]: but then effPerf will be 0, so we are actually fine?

Comment (vyzo) [reply chain/messagepool/selection.go:815]: undefined, good thing you spotted.

**Pull request merged**